core/domain/topic_domain.py

```diff
@@ -601,47 +612,82 @@ def update_subtopic_title(self, subtopic_id, new_title):
+                    % skill_id)
+
+        if old_subtopic_id is None:
+            if skill_id not in self.uncategorized_skill_ids:
+                raise Exception(
+                    'Skill id %s is not an uncategorized skill id.' % skill_id)
+
+        new_subtopic_index = self.get_subtopic_index(new_subtopic_id)
+        if new_subtopic_index is None:
+            raise Exception(
+                'The subtopic with id %s does not exist.' % new_subtopic_id)
+        if skill_id in self.subtopics[new_subtopic_index].skill_ids:
+            raise Exception(
+                'Skill id %s is already present in the target subtopic'
+                % skill_id)
+
+        if old_subtopic_id is None:
+            self.uncategorized_skill_ids.remove(skill_id)
+        else:
+            self.subtopics[old_subtopic_index].skill_ids.remove(skill_id)
+
+        self.subtopics[new_subtopic_index].skill_ids.append(skill_id)
+
+    def remove_skill_from_subtopic(self, subtopic_id, skill_id):
+        """Removes the skill_id from a subtopic and adds it to
+        uncategorized skill ids.
+
+        Args:
+            subtopic_id: str. The subtopic from which the skill is
+                to be removed.
+            skill_id: str. The skill id which is to be removed.
+
+        Raises:
+            Exception. The subtopic with the given id doesn't exist.
+            Exception. The skill id should be present in the old subtopic
+                already before moving.
+        """
+
         subtopic_index = self.get_subtopic_index(subtopic_id)
         if subtopic_index is None:
             raise Exception(
                 'The subtopic with id %s does not exist.' % subtopic_id)
-        is_new_skill_id_present = False
-
-        for subtopic in self.subtopics:
-            if new_skill_id in subtopic.skill_ids:
-                if subtopic.id == subtopic_id:
-                    raise Exception(
-                        'The skill id is already present in the subtopic.')
-                if is_new_skill_id_present:
-                    raise Exception(
-                        'More than 1 subtopic have the same skill with id %s.'
-                        % new_skill_id)
-                subtopic.skill_ids.remove(new_skill_id)
-                is_new_skill_id_present = True
-
-        if new_skill_id in self.uncategorized_skill_ids:
-            self.uncategorized_skill_ids.remove(new_skill_id)
-            is_new_skill_id_present = True
-
-        if not is_new_skill_id_present:
+        if skill_id not in self.subtopics[subtopic_index].skill_ids:
             raise Exception(
-                'The skill id should be present in the topic before moving.')
-        self.subtopics[subtopic_index].skill_ids.append(new_skill_id)
+                'Skill id %s is not present in the old subtopic'
+                % skill_id)
+
+        self.subtopics[subtopic_index].skill_ids.remove(skill_id)
+        if skill_id not in self.uncategorized_skill_ids:
```

Comment:
Why this check? Should be more like an assertion -- we don't expect it to be there already, right?

Reply:
Yeah, there is already a check in add_uncategorized_skill_id() whether a new skill id is present anywhere in the topic, so duplicates won't be there. Will remove this.

Reply:
Done.